source/distributed/cell_weights.cc

```diff
@@ -0,0 +1,186 @@
+    void
+    CellWeights<dim, spacedim>::register_custom_weighting(
+      const std::function<unsigned int(
+        const FiniteElement<dim, spacedim> &,
+        const typename hp::DoFHandler<dim, spacedim>::cell_iterator &)>
+        custom_function)
+    {
+      weighting_function = custom_function;
+    }
+
+
+
+    template <int dim, int spacedim>
+    unsigned int
+    CellWeights<dim, spacedim>::weight_callback(
+      const typename Triangulation<dim, spacedim>::cell_iterator &cell_,
+      const typename Triangulation<dim, spacedim>::CellStatus     status)
+    {
+      // Check if we are still working with the correct combination of
+      // Triangulation and DoFHandler.
+      Assert(&(*triangulation) == &(dof_handler->get_triangulation()),
+             ExcMessage(
+               "Triangulation associated with the DoFHandler has changed!"));
+
+      // Convert cell type from Triangulation to DoFHandler to be able
+      // to access the information about the degrees of freedom.
+      const typename hp::DoFHandler<dim, spacedim>::cell_iterator cell(
+        *cell_, dof_handler);
+
+      // Determine which FiniteElement object will be present on this cell after
+      // refinement and will thus specify the number of degrees of freedom.
+      unsigned int fe_index = numbers::invalid_unsigned_int;
+      switch (status)
+        {
+          case parallel::distributed::Triangulation<dim,
+                                                    spacedim>::CELL_PERSIST:
+          case parallel::distributed::Triangulation<dim, spacedim>::CELL_REFINE:
+          case parallel::distributed::Triangulation<dim,
+                                                    spacedim>::CELL_INVALID:
+            fe_index = cell->active_fe_index();
+            break;
+
+          case parallel::distributed::Triangulation<dim,
+                                                    spacedim>::CELL_COARSEN:
+            {
+              std::set<unsigned int> fe_indices_children;
+              for (unsigned int child_index = 0;
+                   child_index < GeometryInfo<dim>::max_children_per_cell;
+                   ++child_index)
+                fe_indices_children.insert(
+                  cell->child(child_index)->active_fe_index());
+
+              fe_index = dof_handler->get_fe().find_least_face_dominating_fe(
+                fe_indices_children);
+
+              Assert(fe_index != numbers::invalid_unsigned_int,
+                     ExcMessage(
+                       "No FiniteElement has been found in your FECollection "
+                       "that dominates all children of a cell you are trying "
+                       "to coarsen!"));
+            }
+            break;
+
+          default:
+            Assert(false, ExcInternalError());
+            break;
+        }
+
+      // Return the cell weight determined by the function of choice.
+      return weighting_function(dof_handler->get_fe(fe_index), cell);
```

Comment:
in the documentation of the custom weight register function, you will want to explain how exactly the two arguments to the custom weight are determined and mean